src/stream/src/executor/upstream_sink_union.rs

```diff
@@ -311,25 +425,76 @@ impl UpstreamSinkUnionExecutor {
                 "UpstreamSinkUnion",
             ]);
 
-        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align), None);
+        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align.clone()), None);
         pin_mut!(upstreams);
 
-        while let Some(msg) = upstreams.next().await {
-            yield msg?;
+        let mut buffered_barriers = VecDeque::new();
+
+        // Here, `tokio::select` cannot be used directly in `try_stream` function.
+        // Err for breaking the loop. Ok(None) for continuing the loop.
+        let mut select_once = async || -> StreamExecutorResult<Message> {
+            loop {
+                tokio::select! {
+                    biased;
+
+                    // If None is returned, it means upstreams is empty, which
+                    // means we should continue pending and wait on the second branch.
+                    msg = pending_on_none(upstreams.next()) => {
+                        let msg = msg.context("UpstreamSinkUnionExecutor pull upstream failed")?;
+                        if let Message::Barrier(barrier) = &msg {
+                            let current_barrier = buffered_barriers.pop_front().unwrap();
+                            assert_equal_dispatcher_barrier(&current_barrier, barrier);
+                            self.handle_update(&mut upstreams, &barrier_align, barrier).await?;
+                        }
+                        return Ok(msg);
+                    }
+
+                    barrier = self.barrier_rx.recv() => {
```

Comment:
If this would further simplify the implementation: Is it possible to treat the barrier receiver as a special input of `DynamicReceivers`, so that it will never be empty and we don't need extra logic for `select`ing here.